sys/dev/tagscan/tagscan.c

```diff
@@ -0,0 +1,809 @@
+/*-
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) 2025 Robert N. M. Watson
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer
+ *    in this position and unchanged.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <sys/param.h>
+#include <sys/systm.h>
+#include <sys/conf.h>
+#include <sys/kernel.h>
+#include <sys/malloc.h>
+#include <sys/module.h>
+#include <sys/sysctl.h>
+
+#include <vm/vm.h>
+#include <vm/vm_object.h>
+#include <vm/vm_page.h>
+#include <vm/vm_param.h>
+
+#include "tagscan.h"
+
+struct tagscan_stats ts;
+
+static void
+tagscan_page_process(vm_page_t m)
+{
+	vm_object_t obj;
+	uintcap_t c, *cp;
+	vm_paddr_t pa;
+	u_int i, tags, seals, readables, writables;
+
+	tags = seals = readables = writables = 0;
+
+	/*
+	 * Skip pages that don't correspond to actual resources.
+	 */
+	pa = VM_PAGE_TO_PHYS(m);
+	if (pa == 0) {
+		ts.ts_without_pa_count++;
+		return;
+	}
+
+	/*
+	 * Handle counters purely related to the page-level metadata.
+	 */
+	if (m->oflags & VPO_UNMANAGED)
+		ts.ts_unmanaged_count++;
+	if (m->a.flags & PGA_REFERENCED)
+		ts.ts_referenced_count++;
+	if (m->a.flags & PGA_CAPSTORE)
+		ts.ts_capstore_count++;
+	if (m->a.flags & PGA_CAPDIRTY)
+		ts.ts_capdirty_count++;
+	if (m->flags & PG_PCPU_CACHE)
+		ts.ts_pcpu_count++;
+	if (m->flags & PG_FICTITIOUS)
+		ts.ts_fict_count++;
+	if (m->flags & PG_ZERO)
+		ts.ts_zero_count++;
+	if (m->flags & PG_NODUMP)
+		ts.ts_nodump_count++;
+	if (m->flags & PG_NOFREE)
+		ts.ts_nofree_count++;
```

Comment:
Most of these flags are meaningless if the page is free, they might be left set (flags are initialized at page allocation time) and so incorrectly contribute to totals.

That is, I suspect we should verify that `m->order == VM_NFREEORDER` first to ensure that the page isn't enqueued in the physical memory allocator.

Reply:
On the whole the stats we care about are those associated with pages that had an object, and our various overhead / density calculations are based only on those pages. That said, I think we do care about free pages that have tags in them, for some purposes, and we should report this in a more clear way.

What is the authoritative check for m “free” that we should use?

Reply:
That's fair. My comment really only meant to refer to all of these `m->[ao]flags` checks.

The authoritative check is `m->order < VM_NFREEORDER`. If this condition is true, the page is free, else we must have `m->order == VM_NFREEORDER`.